SPEC.md

```diff
@@ -276,41 +276,51 @@ Note that the circuit pseudocode in this specification does not describe zk-SNAR
 
 #### The state root transition proof circuit
 
-##### Inputs
+This circuit proves the correctness of each state root transition.
 
-| Pseudocode name | zk-SNARK input type | Description | Set by |
-|-|-|-|-|
-| `coordinatorPubKey` | Public | The coordinator's public key | Contract |
-| `currentStateRoot` | Public | The current state root | Contract |
-| `msgTreeRoot` | Public | The Merkle root of the message tree | Contract |
-| `msgTreePathIndex` | Public | The Merkle path index of the message in the message tree | Contract |
-| `maxStateLeafIndex` | Public | The maximum leaf index of the state tree | Contract |
-| `userCurrentLeafPathIndex` | Public | The Merkle path index from the user's latest valid state leaf to the current state root | Contract |
-| `newStateRoot` | Public | The new state root | Contract |
-| `userCurrentLeafPathElements` | Private | The Merkle path elements from the user's latest valid state leaf to the current state root | Coordinator |
-| `currentVoteOptionPathElements[n]` | Private <!--`253` * vote option tree depth--> | The Merkle path needed to prove the existence of the current vote option leaf. | Coordinator |
-| `newVoteOptionPathElements[n]` | Private <!--`253` * vote option tree depth--> | The Merkle path needed to update the vote option tree root in the state leaf. | Coordinator |
-| `currentVoteWeight` | Private <!--32--> | In the quadratic voting use case, this is the square root of the number of voice credits a user wishes to spend on this vote. | Coordinator |
-| `message` | Private | The message | Coordinator |
-| `msgTreePathElements` | Private | The Merkle path elements to the message tree root from the message leaf | Coordinator |
-| `randomLeaf` | Private | Random data | Coordinator |
-| `newStateTreePathIndex` | Private |The Merkle path index to the new state root from the new state leaf | Coordinator |
-| `newStateTreePathElements` | Private |The Merkle path elements to the new state root from the new state leaf | Coordinator |
-| `newStateTreePathElementsToZero` | Private |The Merkle path elements to the new state root from leaf 0, **after** the new state leaf has been updated | Coordinator |
-| `userCurrentLeaf` | Private | The user's latest valid state leaf | Coordinator |
-| `command` | Private | The command to process. Includes all the details in the leaf. | Coordinator |
-| `noOp` | Private | The no-op flag | Coordinator |
-| `userPubKey` | Private | The user's public key | Coordinator |
-| `encPubKey` | Private | The ephermeral public key | Contract |
```

Comment:
I was confused by this. If I understand correctly,  `userPubKey` is published by the user via contract, collected by the coordinator to generate a shared key. The `encPubKey` should be the coordinator's private key that fed into the circuit as a private message.

Reply:
To encrypt the command, the user creates an ephemeral keypair, and generates a shared key using `encPrivKey` and `coordinatorPubKey`. The coordinator can recreate this key using `encPubKey` and `coordinatorPrivKey`.

The user's ephemeral keypair's public key is different from the one which they store in the state tree.